File: projects/fibonacci/Py/fib.py
#!/usr/bin/env python3

import time

def fib_rec(n):
    if (n < 0):
        return 0
    if (n < 2):
        return n
    return fib_rec(n-2) + fib_rec(n-1)

def fib_lin(n):
    if (n < 0):
        return 0
    if (n < 2):
        return n
    s0 = 0
    s1 = 1
    s2 = 1
    for i in range(2, n):
        s0 = s1
        s1 = s2
        s2 = s1 + s0
    return s2

N = int(input("Enter an integer: "))

print("Recursive algorithm:")
print("{:>12} {:>12} {:>12}".format("Integer", "Fibonacci", "msecs"))
for i in range(1, N+1):
    t0 = time.time()
    f = fib_rec(i)
    t1 = time.time()
    print("{:12} {:12} {:12.6f}".format(i, f, (t1 - t0) * 1000))

print("Linear algorithm:")
print("{:>12} {:>12} {:>12}".format("Integer", "Fibonacci", "msecs"))
for i in range(1, N+1):
    t0 = time.time()
    f = fib_lin(i)
    t1 = time.time()
    print("{:12} {:12} {:12.6f}".format(i, f, (t1 - t0) * 1000))

File: projects/fibonacci/Go/fib.go
package main

import (
	"fmt"
	"time"
)

func FibRec(n uint) (uint) {
	if n < 2 {
		return n
	}
	return FibRec(n - 2) + FibRec(n - 1)
}

func FibLin(n uint) (uint) {
	var i uint
	var s0, s1, s2 uint

	if n < 2 {
		return n
	}
	s0 = 0
	s1 = 1
	s2 = 1
	for i = 2; i < n; i++ {
		s0 = s1
		s1 = s2
		s2 = s0 + s1
	}
	return s2
}

func main() {
	var i, N uint
	var err error

	fmt.Print("Enter an integer: ")
	_, err = fmt.Scanf("%d", &N);
	if err != nil {
		panic(err)
	}
	fmt.Printf("Recursive algorithm:\n")
	fmt.Printf("%12s %12s %12s\n", "Integer", "Fibonacci", "msecs")
	for i = 1; i <= N; i++ {
		t0 := time.Now()
		f := FibRec(i)
		t1 := time.Now()
		fmt.Printf("%12v %12v %12.6f\n",
			   i, f, t1.Sub(t0).Seconds() * 1000)
	}
	fmt.Printf("Linear algorithm:\n")
	fmt.Printf("%12s %12s %12s\n", "Integer", "Fibonacci", "msecs")
	for i = 1; i <= N; i++ {
		t0 := time.Now()
		f := FibLin(i)
		t1 := time.Now()
		fmt.Printf("%12v %12v %12.6f\n",
			   i, f, t1.Sub(t0).Seconds() * 1000)
	}
}
